In [115]:
import numpy as np 

Exercise 1 : standard deviation 


In [62]:
mean = (8+12+16+24+4)/5
print (mean)

12.8


Variance = (8 - 12.8)^2 + (12 - 12.8)^2 + (16 - 12.8)^2 + (24 - 12.8)^2 + (4 - 12.8)^2 = 47.36

Standard deviation = Square root of Variance = 6.88 (approx)

In [63]:
data = np.array([8, 12, 16, 24, 4])

variance = np.var(data)

std_dev = np.std(data)

variance, std_dev

(47.36, 6.881860213634101)

The sum of deviations will be 0 because the deviations below the mean cancel the deviations above the mean 

Exercise 2: 

Introducing outliers to a dataset typically increases the standard deviation. The standard deviation measures the amount of variation or dispersion in a set of values. Outliers are data points that lie far from the other points in the dataset, thus significantly affecting the overall variance. Since the standard deviation is the square root of the variance, when variance increases due to outliers, the standard deviation also increases. This is because outliers will increase the average of the squared deviations from the mean, which is the basis for calculating variance and, consequently, the standard deviation.

Exercise 3: Answer F, both B and C

Exercise 4: 13 could be the value of the median.

In [64]:
list_4 = [10, 10, 10, 10, 11, 11, 12, 13, 14 ]
[15, 15, 15, 16, 16, 16, 16, 16, 17, 18, 19, 19]
[20, 20, 20, 21, 21, 21, 21, 21, 22, 22, 22, 23, 23, 23, 24, 24, 24]
[26, 26, 26, 26, 26, 26, 26, 27, 27, 27, 27, 29, 29, 29, 29, 29, 29, 29, 29, 29]
[30, 30, 30, 30, 30, 31, 31, 32, 32, 33, 33, 34, 34]
[35, 35, 35, 37, 37, 37, 37, 39, 39, 39]
[40, 40, 43, 43, 44, 44]
[46, 46, 46, 47, 49]
[52, 53, 54]

median_4 = np.median(list_4)
median_4

11.0

Exercise 5: Postive Skew or skewed right 

Exercise 6: THe measure at the center of the data could be the mean.

Exercise 7: according to the graph, 35 have a salary of less than 10k.

In [65]:
#Exercise 8: 

list_8 = np.array([12, 15, 14, 15, 6, 12, 19, 3])



In [66]:
mean = np.mean(list_8)
print (mean)

12.0


In [67]:
median = np.median(list_8)
print(median)

13.0


Exercise 9: 

In [68]:
arr = np.array([[25,3],[40,2],[35,1],[50,1]])

In [69]:
std_dev = np.std(arr)
std_dev

18.98642607232862

Exercise 10
A discrete random variable is one that can take on a countable number of distinct values. 

B) the number of pens purchased by a student in a year


Exercice 11

In [70]:
A) mu=5, sigma =1.5
B) mu=10, sigma =1.0
C) mu=5, sigma =1.65
D) mu=8, sigma =1.2
E) mu=10, sigma =1.6

Variances 

A = 2,24
B = 40,5
C = 5,61
D = 23,12
E = 35,28

#According to the range and the variation, B has the greatest spread. 

SyntaxError: unmatched ')' (2548720994.py, line 1)

Exercices extra 

In [ ]:
import numpy as np

In [88]:
#### 3 
data_set = np.zeros(10)

arr = np.array(data_set)

In [89]:
arr

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [96]:
#### 4 

arr.size

10

In [98]:
arr.shape

(10,)

In [72]:
###5
help(np.add)


Help on ufunc:

add = <ufunc 'add'>
    add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])
    
    Add arguments element-wise.
    
    Parameters
    ----------
    x1, x2 : array_like
        The arrays to be added.
        If ``x1.shape != x2.shape``, they must be broadcastable to a common
        shape (which becomes the shape of the output).
    out : ndarray, None, or tuple of ndarray and None, optional
        A location into which the result is stored. If provided, it must have
        a shape that the inputs broadcast to. If not provided or None,
        a freshly-allocated array is returned. A tuple (possible only as a
        keyword argument) must have length equal to the number of outputs.
    where : array_like, optional
        This condition is broadcast over the input. At locations where the
        condition is True, the `out` array will be set to the ufunc result.
        Elsewhere, the `out` array wi

In [75]:
doc_string = np.add.__doc__
print(doc_string[:500])

add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])

Add arguments element-wise.

Parameters
----------
x1, x2 : array_like
    The arrays to be added.
    If ``x1.shape != x2.shape``, they must be broadcastable to a common
    shape (which becomes the shape of the output).
out : ndarray, None, or tuple of ndarray and None, optional
    A location into which the result is stored. If provided, it must have
    a shape that the input


In [93]:
 ####6

data_set[4]=1

data_set


array([0., 0., 0., 0., 1., 0., 0., 0., 0., 0.])

In [101]:
####7

data_set_7 = np.arange(10,49)

In [102]:
data_set_7

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48])

In [105]:
#### 8

reverse_data_set_8 = np.flip(data_set_7)

reverse_data_set_8

array([48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32,
       31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15,
       14, 13, 12, 11, 10])

In [107]:
#### 9

data_set_9 = np.arange(0,9).reshape(3,3)
print(data_set_9)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


In [110]:
#### 10

np.nonzero([1,2,0,0,4,0]) 



(array([0, 1, 4]),)

In [114]:
#### 11 3x3 Matrix 

data_set_11 = np.identity(3)

print(data_set_11)

#np.identity() function creates a 2D identity matrix with a size of 3x3.
# An identity matrix is a square matrix with ones on the diagonal and zeros 
elsewhere.


[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [116]:
np.eye(3)


array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [122]:
#### 12

list_12 = np.random.random((3, 3, 3))

print(list_12)

[[[0.20863933 0.17328369 0.04283742]
  [0.43370716 0.9180931  0.45746366]
  [0.90700605 0.76486202 0.31290436]]

 [[0.04765308 0.25937664 0.94285296]
  [0.9171585  0.20117648 0.84788891]
  [0.31373064 0.29529087 0.84723191]]

 [[0.92837781 0.81275807 0.27183629]
  [0.09909935 0.41132741 0.93596023]
  [0.99266284 0.84172655 0.44291743]]]


In [125]:
### 13

list_13 = np.random.random((10,10))

list_13

array([[0.19980616, 0.78864001, 0.95654269, 0.024898  , 0.49575444,
        0.05170355, 0.82301248, 0.51467337, 0.01587695, 0.58756539],
       [0.61086855, 0.92161429, 0.43942419, 0.58909069, 0.79879664,
        0.73507236, 0.14711082, 0.41024664, 0.69868868, 0.34513623],
       [0.38406262, 0.79426664, 0.59741195, 0.83544191, 0.62771021,
        0.18227558, 0.56421632, 0.0414889 , 0.92612502, 0.85240187],
       [0.39812473, 0.42755447, 0.75262543, 0.65312126, 0.12892145,
        0.67769722, 0.51181032, 0.88475171, 0.44458691, 0.56069348],
       [0.10354579, 0.99776557, 0.15895998, 0.52187291, 0.26860386,
        0.24867819, 0.55932695, 0.01253963, 0.53265733, 0.94635563],
       [0.93168348, 0.06131074, 0.23823491, 0.59816558, 0.65031526,
        0.04589124, 0.8859214 , 0.0917191 , 0.60466791, 0.95937337],
       [0.08125601, 0.7615407 , 0.57565356, 0.44643435, 0.92404245,
        0.12891863, 0.17443241, 0.71267289, 0.83963301, 0.48451761],
       [0.48134347, 0.76487653, 0.8366539

In [129]:
min_value = list_13.min()
max_value = list_13.max()

min_value


0.006942130569822069

In [130]:
max_value

0.9977655695195004

In [140]:
#### 14

list_14 = np.random.random(30)

list_14

array([0.9621996 , 0.35850558, 0.08949821, 0.40825439, 0.51891041,
       0.1837182 , 0.39761719, 0.14600835, 0.97588448, 0.19388879,
       0.07498523, 0.86130649, 0.08876403, 0.36050689, 0.46571855,
       0.6879825 , 0.16694464, 0.14830316, 0.07424074, 0.45426135,
       0.28657897, 0.35069517, 0.04205424, 0.87930932, 0.13023406,
       0.94925341, 0.97978695, 0.44892947, 0.79537425, 0.52427629])

In [141]:
mean_14 = np.mean(list_14)

mean_14

0.4334663635408535

In [148]:
##### 15

list_15 = np.ones((8, 8))

list_15

array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]])

In [151]:
list_15[1:-1, 1:-1] = 0

list_15

array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]])

In [156]:
#### 16

list_16 = np.ones((3,3))

list_16

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [158]:
list_16_2 = np.pad(list_16, pad_width=1, mode='constant', constant_values=0)

list_16_2

array([[0., 0., 0., 0., 0.],
       [0., 1., 1., 1., 0.],
       [0., 1., 1., 1., 0.],
       [0., 1., 1., 1., 0.],
       [0., 0., 0., 0., 0.]])

In [159]:
### 17

0 * np.nan # equal nan 
np.nan == np.nan    #False 
np.inf > np.nan    #False
np.nan - np.nan   #  Equal nan 
np.nan in set([np.nan])   #True
0.3 == 3 * 0.1  #False 

False

In [164]:
### 18 

list_18 = np.diag([0,1,2,3,4])

list_18

array([[0, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 2, 0, 0],
       [0, 0, 0, 3, 0],
       [0, 0, 0, 0, 4]])

In [176]:
### 19

list_19 = np.zeros((8,8))

list_19

array([[0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0.]])

In [177]:
list_19[1::2, ::2] = 1  # Setting alternate rows starting from the second row with alternate columns to 1
list_19[::2, 1::2] = 1 

list_19

array([[0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.]])

In [181]:
## 20

np.unravel_index (100, (6,7,8))


(1, 5, 4)